README.md
# Alibaba Cloud MNS driver for Laravel Queue

[![Tests](https://github.com/dew-serverless/laravel-mns-driver/actions/workflows/tests.yml/badge.svg)](https://github.com/dew-serverless/laravel-mns-driver/actions/workflows/tests.yml)
[![Latest Version](https://img.shields.io/packagist/v/dew-serverless/laravel-mns-driver)](https://github.com/dew-serverless/laravel-mns-driver)
[![Total Downloads](https://img.shields.io/packagist/dt/dew-serverless/laravel-mns-driver)](https://github.com/dew-serverless/laravel-mns-driver)
[![License](https://img.shields.io/packagist/l/dew-serverless/laravel-mns-driver)](https://github.com/dew-serverless/laravel-mns-driver)

The driver gives the ability to run Laravel Queue on top of ACS MNS, a distributed messaging service. The serverless MNS also relieves developers from the burden of infrastructure maintenance and helps build a loosely coupled system.

Checkout documentation at [Wiki](https://github.com/dew-serverless/laravel-mns-driver/wiki).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.1.0] - 2023-11-10

### Added

- Exposes MNS queue instance by @lizhineng in https://github.com/dew-serverless/laravel-mns-driver/pull/11

## [2.0.0] - 2023-11-07

### Added

- Supports clear queue by @lizhineng in https://github.com/dew-serverless/laravel-mns-driver/pull/8
- Supports customize HTTP client by @lizhineng in https://github.com/dew-serverless/laravel-mns-driver/pull/10
- Introduce PHPStan in the workflow by @lizhineng in https://github.com/dew-serverless/laravel-mns-driver/pull/6

### Changed

- **[BREAKING CHANGES]** Throws exception when result was failed by @lizhineng in https://github.com/dew-serverless/laravel-mns-driver/pull/7
- Improve readability of test cases by @lizhineng in https://github.com/dew-serverless/laravel-mns-driver/pull/9

### Fixed

- Fix connector registration on service provider by @lizhineng in https://github.com/dew-serverless/laravel-mns-driver/pull/4
- Fix connector not passing the correct queue instance by @lizhineng in https://github.com/dew-serverless/laravel-mns-driver/pull/5
- Fix PHPStan rules by @lizhineng in https://github.com/dew-serverless/laravel-mns-driver/pull/7

## [1.0.0] - 2023-11-05

### Added

- First release.
- Hopes the package could help you easily integrating ACS MNS with your Laravel Queue.

[unreleased]: https://github.com/dew-serverless/laravel-mns-driver/compare/v2.1.0...HEAD
[2.1.0]: https://github.com/dew-serverless/laravel-mns-driver/compare/v2.0.0...v2.1.0
[2.0.0]: https://github.com/dew-serverless/laravel-mns-driver/compare/v1.0.0...v2.0.0
[1.0.0]: https://github.com/dew-serverless/laravel-mns-driver/releases/tag/v1.0.0


tests/MnsJobTest.php
<?php

use Dew\Mns\Contracts\XmlEncoder;
use Dew\Mns\MnsClient;
use Dew\Mns\Versions\V20150606\Queue;
use Dew\Mns\Versions\V20150606\Results\ChangeMessageVisibilityResult;
use Dew\Mns\Versions\V20150606\Results\ReceiveMessageResult;
use Dew\Mns\Versions\V20150606\Results\Result;
use Dew\MnsDriver\MnsJob;
use Illuminate\Container\Container;
use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\StreamInterface;

beforeEach(function () {
    $this->client = new MnsClient('http://1234567891011.mns.us-west-1.aliyuncs.com', 'key', 'secret');
    $this->mns = Mockery::mock(new Queue($this->client));
    $this->mockedContainer = Mockery::mock(Container::class);
    $this->queueName = 'default';

    $this->mockedJob = 'greeting';
    $this->mockedData = ['message' => 'Hello world!'];
    $this->mockedPayload = json_encode(['job' => $this->mockedJob, 'data' => $this->mockedData]);
    $this->mockedMessageId = '5F290C926D472878-2-14D9529****-200000001';
    $this->mockedReceiptHandle = '1-ODU4OTkzNDU5My0xNDM1MTk3NjAwLTItNg==';

    $this->mockedStream = Mockery::mock(StreamInterface::class);
    $this->mockedStream->allows()->__toString()->andReturns('<response></response>');

    $this->mockedOkResponse = Mockery::mock(ResponseInterface::class);
    $this->mockedOkResponse->allows()->getStatusCode()->andReturns(200);
    $this->mockedOkResponse->allows()->getHeaderLine('content-type')->andReturns('text/xml');
    $this->mockedOkResponse->allows()->getBody()->andReturns($this->mockedStream);

    $this->mockedNoContentResponse = Mockery::mock(ResponseInterface::class);
    $this->mockedNoContentResponse->allows()->getStatusCode()->andReturns(204);
    $this->mockedNoContentResponse->allows()->getHeaderLine('content-type')->andReturns('');

    $this->mockedReceiveMessageResult = new ReceiveMessageResult($this->mockedOkResponse, tap(Mockery::mock(XmlEncoder::class), function ($mock) {
        $mock->expects()->decode('<response></response>')->andReturns([
            'MessageId' => $this->mockedMessageId,
            'ReceiptHandle' => $this->mockedReceiptHandle,
            'MessageBody' => $this->mockedPayload,
            'MessageBodyMD5' => md5($this->mockedPayload),
        ]);
    }));

    $this->mockedChangeMessageVisibilityResult = new ChangeMessageVisibilityResult($this->mockedOkResponse, tap(Mockery::mock(XmlEncoder::class), function ($mock) {
        $mock->expects()->decode('<response></response>')->andReturns([
            'ReceiptHandle' => $this->mockedReceiptHandle,
        ]);
    }));

    $this->mockedDeleteMessageResult = new Result($this->mockedNoContentResponse, Mockery::mock(XmlEncoder::class));
});

test('release releases the job onto mns', function () {
    $this->mns->expects()->changeMessageVisibility($this->queueName, $this->mockedReceiptHandle, 60)->andReturns($this->mockedChangeMessageVisibilityResult);
    $job = new MnsJob($this->mockedContainer, $this->mns, $this->mockedReceiveMessageResult, 'mns', $this->queueName);
    $job->release(60);
});

test('delete removes the job from mns', function () {
    $this->mns->expects()->deleteMessage($this->queueName, $this->mockedReceiptHandle)->andReturns($this->mockedDeleteMessageResult);
    $job = new MnsJob($this->mockedContainer, $this->mns, $this->mockedReceiveMessageResult, 'mns', $this->queueName);
    $job->delete();
});

test('fire calls the job handler', function () {
    $job = new MnsJob($this->mockedContainer, $this->mns, $this->mockedReceiveMessageResult, 'mns', $this->queueName);
    $job->getContainer()->expects()->make($this->mockedJob)->andReturns($handler = Mockery::mock(stdClass::class));
    $handler->expects()->fire($job, $this->mockedData);
    $job->fire();
});


tests/MnsQueueTest.php
<?php

use Dew\Mns\Contracts\XmlEncoder;
use Dew\Mns\MnsClient;
use Dew\Mns\Versions\V20150606\Queue;
use Dew\Mns\Versions\V20150606\Results\BatchDeleteMessageResult;
use Dew\Mns\Versions\V20150606\Results\BatchReceiveMessageResult;
use Dew\Mns\Versions\V20150606\Results\GetQueueAttributesResult;
use Dew\Mns\Versions\V20150606\Results\ReceiveMessageResult;
use Dew\Mns\Versions\V20150606\Results\SendMessageResult;
use Dew\MnsDriver\MnsJob;
use Dew\MnsDriver\MnsQueue;
use Illuminate\Container\Container;
use Illuminate\Support\Carbon;
use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\StreamInterface;

beforeEach(function () {
    $this->client = new MnsClient('http://1234567891011.mns.us-west-1.aliyuncs.com', 'key', 'secret');
    $this->mns = Mockery::mock(new Queue($this->client));
    $this->queueName = 'default';

    $this->mockedJob = 'greeting';
    $this->mockedData = ['message' => 'Hello world!'];
    $this->mockedPayload = json_encode(['job' => $this->mockedJob, 'data' => $this->mockedData]);
    $this->mockedDelay = 60;
    $this->mockedMessageId = '5F290C926D472878-2-14D9529****-200000001';
    $this->mockedReceiptHandle = '1-ODU4OTkzNDU5My0xNDM1MTk3NjAwLTItNg==';

    $this->mockedActiveMessages = 3;
    $this->mockedInactiveMessages = 4;
    $this->mockedDelayMessages = 5;

    $this->mockedStream = Mockery::mock(StreamInterface::class);
    $this->mockedStream->allows()->__toString()->andReturns('<response></response>');
    $this->mockedOkResponse = Mockery::mock(ResponseInterface::class);
    $this->mockedOkResponse->allows()->getStatusCode()->andReturns(200);
    $this->mockedOkResponse->allows()->getHeaderLine('content-type')->andReturns('text/xml');
    $this->mockedOkResponse->allows()->getBody()->andReturns($this->mockedStream);

    $this->mockedCreatedResponse = Mockery::mock(ResponseInterface::class);
    $this->mockedCreatedResponse->allows()->getStatusCode()->andReturns(201);
    $this->mockedCreatedResponse->allows()->getHeaderLine('content-type')->andReturns('text/xml');
    $this->mockedCreatedResponse->allows()->getBody()->andReturns($this->mockedStream);

    $this->mockedNoContentResponse = Mockery::mock(ResponseInterface::class);
    $this->mockedNoContentResponse->allows()->getStatusCode()->andReturns(204);
    $this->mockedNoContentResponse->allows()->getHeaderLine('content-type')->andReturns('');

    $this->mockedNotFoundResponse = Mockery::mock(ResponseInterface::class);
    $this->mockedNotFoundResponse->allows()->getStatusCode()->andReturn(404);
    $this->mockedNotFoundResponse->allows()->getHeaderLine('content-type')->andReturns('text/xml');
    $this->mockedNotFoundResponse->allows()->getBody()->andReturns($this->mockedStream);

    $this->mockedGetQueueAttributesResult = new GetQueueAttributesResult($this->mockedOkResponse, tap(Mockery::mock(XmlEncoder::class), function ($mock) {
        $mock->expects()->decode('<response></response>')->andReturns([
            'ActiveMessages' => $this->mockedActiveMessages,
            'InactiveMessages' => $this->mockedInactiveMessages,
            'DelayMessages' => $this->mockedDelayMessages,
        ]);
    }));

    $this->mockedSendMessageResult = new SendMessageResult($this->mockedCreatedResponse, tap(Mockery::mock(XmlEncoder::class), function ($mock) {
        $mock->expects()->decode('<response></response>')->andReturns([
            'MessageId' => $this->mockedMessageId,
            'MessageBodyMD5' => md5($this->mockedPayload),
            'ReceiptHandle' => $this->mockedReceiptHandle,
        ]);
    }));

    $this->mockedReceiveMessageResult = new ReceiveMessageResult($this->mockedOkResponse, tap(Mockery::mock(XmlEncoder::class), function ($mock) {
        $mock->expects()->decode('<response></response>')->andReturns([
            'MessageId' => $this->mockedMessageId,
            'MessageBody' => $this->mockedPayload,
            'MessageBodyMD5' => md5($this->mockedPayload),
            'ReceiptHandle' => $this->mockedReceiptHandle,
        ]);
    }));

    $this->mockedMessageNotExistsResult = new ReceiveMessageResult($this->mockedNotFoundResponse, tap(Mockery::mock(XmlEncoder::class), function ($mock) {
        $mock->expects()->decode('<response></response>')->andReturns([
            'Code' => 'MessageNotExist',
            'Message' => 'Message not exist.',
        ]);
    }));

    $this->mockedBatchReceiveMessageResult = new BatchReceiveMessageResult($this->mockedOkResponse, tap(Mockery::mock(XmlEncoder::class), function ($mock) {
        $mock->expects()->decode('<response></response>')->andReturns([
            'Message' => [
                ['ReceiptHandle' => $this->mockedReceiptHandle],
            ],
        ]);
    }));

    $this->mockedBatchMessageNotFoundResult = new BatchReceiveMessageResult($this->mockedNotFoundResponse, tap(Mockery::mock(XmlEncoder::class), function ($mock) {
        $mock->expects()->decode('<response></response>')->andReturns([
            'Code' => 'MessageNotExist',
            'Message' => 'Message not exist.',
        ]);
    }));

    $this->mockedBatchDeleteMessageResult = new BatchDeleteMessageResult($this->mockedNoContentResponse, Mockery::mock(XmlEncoder::class));
});

test('queue size includes active, inactive and delayed messages', function () {
    $this->mns->expects()->getQueueAttributes($this->queueName)->andReturns($this->mockedGetQueueAttributesResult);
    $queue = new MnsQueue($this->mns, $this->queueName);
    expect($queue->size())->toBeInt()->toBe($this->mockedActiveMessages + $this->mockedInactiveMessages + $this->mockedDelayMessages);
});

test('push job to mns', function () {
    $queue = $this->getMockBuilder(MnsQueue::class)->onlyMethods(['createPayload'])->setConstructorArgs([$this->mns, $this->queueName])->getMock();
    $queue->setContainer($container = Mockery::spy(Container::class));
    $queue->expects($this->once())->method('createPayload')->with($this->mockedJob, $this->queueName, $this->mockedData)->willReturn($this->mockedPayload);
    $this->mns->expects()->sendMessage($this->queueName, ['MessageBody' => $this->mockedPayload])->andReturns($this->mockedSendMessageResult);
    $id = $queue->push($this->mockedJob, $this->mockedData, $this->queueName);
    expect($id)->toBe($this->mockedMessageId);
    $container->shouldHaveReceived('bound')->with('events')->once();
});

test('push delayed job to mns', function () {
    $queue = $this->getMockBuilder(MnsQueue::class)->onlyMethods(['createPayload'])->setConstructorArgs([$this->mns, $this->queueName])->getMock();
    $queue->setContainer($container = Mockery::spy(Container::class));
    $queue->expects($this->once())->method('createPayload')->with($this->mockedJob, $this->queueName, $this->mockedData)->willReturn($this->mockedPayload);
    $this->mns->expects()->sendMessage($this->queueName, ['MessageBody' => $this->mockedPayload, 'DelaySeconds' => $this->mockedDelay])->andReturns($this->mockedSendMessageResult);
    $id = $queue->later($this->mockedDelay, $this->mockedJob, $this->mockedData, $this->queueName);
    expect($id)->toBe($this->mockedMessageId);
    $container->shouldHaveReceived('bound')->with('events')->once();
});

test('push delayed job with datetime to mns', function () {
    $now = Carbon::now();
    $queue = $this->getMockBuilder(MnsQueue::class)->onlyMethods(['createPayload'])->setConstructorArgs([$this->mns, $this->queueName])->getMock();
    $queue->setContainer($container = Mockery::spy(Container::class));
    $queue->expects($this->once())->method('createPayload')->with($this->mockedJob, $this->queueName, $this->mockedData)->willReturn($this->mockedPayload);
    $this->mns->expects()->sendMessage($this->queueName, ['MessageBody' => $this->mockedPayload, 'DelaySeconds' => $this->mockedDelay])->andReturns($this->mockedSendMessageResult);
    $id = $queue->later($now->addSeconds($this->mockedDelay), $this->mockedJob, $this->mockedData, $this->queueName);
    expect($id)->toBe($this->mockedMessageId);
    $container->shouldHaveReceived('bound')->with('events')->once();
});

test('pop job off of mns', function () {
    $queue = new MnsQueue($this->mns, $this->queueName);
    $queue->setContainer(Mockery::mock(Container::class));
    $queue->setConnectionName('mns');
    $this->mns->expects()->receiveMessage($this->queueName)->andReturns($this->mockedReceiveMessageResult);
    $job = $queue->pop($this->queueName);
    expect($job)->toBeInstanceOf(MnsJob::class);
});

test('pop job off of empty mns', function () {
    $queue = new MnsQueue($this->mns, $this->queueName);
    $this->mns->expects()->receiveMessage($this->queueName)->andReturns($this->mockedMessageNotExistsResult);
    $job = $queue->pop($this->queueName);
    expect($job)->toBe(null);
});

test('clear queue', function () {
    $queue = new MnsQueue($this->mns, $this->queueName);
    $this->mns->allows()->batchReceiveMessage($this->queueName, ['numOfMessages' => '16', 'waitseconds' => '30'])->andReturns($this->mockedBatchReceiveMessageResult, $this->mockedBatchMessageNotFoundResult);
    $this->mns->expects()->batchDeleteMessage($this->queueName, [$this->mockedReceiptHandle])->andReturns($this->mockedBatchDeleteMessageResult);
    $deleted = $queue->clear($this->queueName);
    expect($deleted)->toBe(1);
});

test('clear empty queue', function () {
    $queue = new MnsQueue($this->mns, $this->queueName);
    $this->mns->expects()->batchReceiveMessage($this->queueName, ['numOfMessages' => '16', 'waitseconds' => '30'])->andReturns($this->mockedBatchMessageNotFoundResult);
    $deleted = $queue->clear($this->queueName);
    expect($deleted)->toBe(0);
});


src/MnsJob.php
<?php

namespace Dew\MnsDriver;

use Dew\Mns\Versions\V20150606\Queue;
use Dew\Mns\Versions\V20150606\Results\ReceiveMessageResult;
use Illuminate\Container\Container;
use Illuminate\Contracts\Queue\Job as JobContract;
use Illuminate\Queue\Jobs\Job;

class MnsJob extends Job implements JobContract
{
    /**
     * Create a new job instance.
     *
     * @param  \Dew\Mns\Versions\V20150606\Queue  $mns
     */
    public function __construct(
        Container $container,
        protected $mns,
        protected ReceiveMessageResult $job,
        string $connectionName,
        string $queue
    ) {
        $this->container = $container;
        $this->connectionName = $connectionName;
        $this->queue = $queue;
    }

    /**
     * Release the job back into the queue after (n) seconds.
     *
     * @param  int  $delay
     * @return void
     */
    public function release($delay = 0)
    {
        parent::release($delay);

        $result = $this->mns->changeMessageVisibility(
            $this->queue, (string) $this->job->receiptHandle(), $delay
        );

        if ($result->failed()) {
            throw new MnsQueueException(sprintf('Release the job with error [%s] %s',
                $result->errorCode(), $result->errorMessage()
            ));
        }
    }

    /**
     * Delete the job from the queue.
     *
     * @return void
     */
    public function delete()
    {
        parent::delete();

        $result = $this->mns->deleteMessage($this->queue, (string) $this->job->receiptHandle());

        if ($result->failed()) {
            throw new MnsQueueException(sprintf('Delete the job with error [%s] %s',
                $result->errorCode(), $result->errorMessage()
            ));
        }
    }

    /**
     * Get the number of times the job has been attempted.
     *
     * @return int
     */
    public function attempts()
    {
        return (int) $this->job->dequeueCount();
    }

    /**
     * Get the job identifier.
     *
     * @return string
     */
    public function getJobId()
    {
        return (string) $this->job->messageId();
    }

    /**
     * Get the raw body string for the job.
     *
     * @return string
     */
    public function getRawBody()
    {
        return (string) $this->job->messageBody();
    }

    /**
     * Get the underlying MNS queue instance.
     */
    public function getMns(): Queue
    {
        return $this->mns;
    }

    /**
     * Get the underlying raw MNS job.
     */
    public function getMnsJob(): ReceiveMessageResult
    {
        return $this->job;
    }
}


src/MnsConnector.php
<?php

namespace Dew\MnsDriver;

use Dew\Mns\MnsClient;
use Dew\Mns\Versions\V20150606\Queue;
use Illuminate\Queue\Connectors\ConnectorInterface;

/**
 * @phpstan-type Config array{
 *     endpoint: string,
 *     key: string,
 *     secret: string,
 *     queue: string,
 *     http?: array<string, mixed>
 * }
 */
class MnsConnector implements ConnectorInterface
{
    /**
     * Establish a queue connection.
     *
     * @param  array<mixed>  $config
     * @return \Illuminate\Contracts\Queue\Queue
     */
    public function connect(array $config)
    {
        /** @var Config $config */
        $mns = new MnsClient($config['endpoint'], $config['key'], $config['secret']);

        $mns->configure($this->withDefaultConfiguration($config['http'] ?? []));

        return new MnsQueue(new Queue($mns), $config['queue']);
    }

    /**
     * Build a configuration with default one.
     *
     * @param  array<string, mixed>  $config
     * @return array<string, mixed>
     */
    protected function withDefaultConfiguration(array $config = []): array
    {
        // timeout: receiving messages could take up to 30 seconds.
        return array_merge([
            'timeout' => 60.0,
        ], $config);
    }
}


src/MnsQueue.php
<?php

namespace Dew\MnsDriver;

use Dew\Mns\Versions\V20150606\Models\Message;
use Dew\Mns\Versions\V20150606\Results\BatchReceiveMessageResult;
use Illuminate\Contracts\Queue\ClearableQueue;
use Illuminate\Contracts\Queue\Queue as QueueContract;
use Illuminate\Queue\Queue;

class MnsQueue extends Queue implements ClearableQueue, QueueContract
{
    /**
     * Create a new MNS queue instance.
     *
     * @param  \Dew\Mns\Versions\V20150606\Queue  $mns
     */
    public function __construct(
        protected $mns,
        protected string $default
    ) {
        //
    }

    /**
     * Get the size of the queue.
     *
     * @param  string|null  $queue
     * @return int
     */
    public function size($queue = null)
    {
        $result = $this->mns->getQueueAttributes($this->getQueue($queue));

        if ($result->failed()) {
            throw new MnsQueueException(sprintf('Get the size of the queue with error [%s] %s',
                $result->errorCode(), $result->errorMessage()
            ));
        }

        return (int) $result->activeMessages()
            + (int) $result->inactiveMessages()
            + (int) $result->delayMessages();
    }

    /**
     * Push a new job onto the queue.
     *
     * @param  string|object  $job
     * @param  mixed  $data
     * @param  string|null  $queue
     * @return mixed
     */
    public function push($job, $data = '', $queue = null)
    {
        $queue = $this->getQueue($queue);

        return $this->enqueueUsing(
            $job,
            $this->createPayload($job, $queue, $data),
            $queue,
            null,
            fn ($payload, $queue) => $this->pushRaw($payload, $queue)
        );
    }

    /**
     * Push a raw payload onto the queue.
     *
     * @param  string  $payload
     * @param  string|null  $queue
     * @param  array<mixed>  $options
     * @return mixed
     */
    public function pushRaw($payload, $queue = null, array $options = [])
    {
        $result = $this->mns->sendMessage($this->getQueue($queue), array_merge([
            'MessageBody' => $payload,
        ], $options));

        if ($result->failed()) {
            throw new MnsQueueException(sprintf('Push job to queue with error [%s] %s',
                $result->errorCode(), $result->errorMessage()
            ));
        }

        return $result->messageId();
    }

    /**
     * Push a new job onto the queue after (n) seconds.
     *
     * @param  \DateTimeInterface|\DateInterval|int  $delay
     * @param  string|object  $job
     * @param  mixed  $data
     * @param  string|null  $queue
     * @return mixed
     */
    public function later($delay, $job, $data = '', $queue = null)
    {
        $queue = $this->getQueue($queue);

        return $this->enqueueUsing(
            $job,
            $this->createPayload($job, $queue, $data),
            $queue,
            $delay,
            fn ($payload, $queue, $delay) => $this->pushRaw($payload, $queue, [
                'DelaySeconds' => $this->secondsUntil($delay),
            ])
        );
    }

    /**
     * Push an array of jobs onto the queue.
     *
     * @param  array<mixed>  $jobs
     * @param  mixed  $data
     * @param  string|null  $queue
     * @return void
     */
    public function bulk($jobs, $data = '', $queue = null)
    {
        foreach ((array) $jobs as $job) {
            if (is_object($job) && isset($job->delay)) {
                $this->later($job->delay, $job, $data, $queue);
            } else {
                /** @var object|string $job */
                $this->push($job, $data, $queue);
            }
        }
    }

    /**
     * Pop the next job off of the queue.
     *
     * @param  string|null  $queue
     * @return \Illuminate\Contracts\Queue\Job|null
     */
    public function pop($queue = null)
    {
        $result = $this->mns->receiveMessage($queue = $this->getQueue($queue));

        if ($result->failed() && $result->errorCode() === 'MessageNotExist') {
            return null;
        }

        if ($result->failed()) {
            throw new MnsQueueException(sprintf('Pop job off of queue with error [%s] %s',
                $result->errorCode(), $result->errorMessage()
            ));
        }

        return new MnsJob(
            $this->container, $this->mns, $result,
            $this->connectionName, $queue
        );
    }

    /**
     * Delete all of the jobs from the queue.
     *
     * @param  string  $queue
     * @return int
     */
    public function clear($queue)
    {
        $deleted = 0;

        $queue = $this->getQueue($queue);

        while (true) {
            $result = $this->retrieveMessages($queue);

            if (! $result instanceof BatchReceiveMessageResult) {
                break;
            }

            /** @var \Dew\Mns\Versions\V20150606\Models\Message[] */
            $messages = $result->messages();

            $receipts = array_filter(
                array_map(fn (Message $message) => $message->receiptHandle(), $messages)
            );

            $deleted += $this->deleteMessages($queue, $receipts);
        }

        return $deleted;
    }

    /**
     * Retrieve the messages from queue.
     */
    protected function retrieveMessages(string $queue): BatchReceiveMessageResult|bool
    {
        $result = $this->mns->batchReceiveMessage($queue, [
            'numOfMessages' => '16',  // max messages we could get at a time
            'waitseconds' => '30',    // max timeout
        ]);

        if ($result->failed() && $result->errorCode() === 'MessageNotExist') {
            return false;
        }

        if ($result->failed()) {
            throw new MnsQueueException(sprintf('Retrieve the messages with error [%s] %s',
                $result->errorCode(), $result->errorMessage()
            ));
        }

        return $result;
    }

    /**
     * Delete the messages from queue.
     *
     * @param  array<int, string>  $receipts
     */
    protected function deleteMessages(string $queue, array $receipts): int
    {
        $result = $this->mns->batchDeleteMessage($queue, $receipts);

        if ($result->failed()) {
            throw new MnsQueueException(sprintf('Delete the messages with error [%s] %s',
                $result->errorCode(), $result->errorMessage()
            ));
        }

        $errors = $result->errors();

        if (is_array($errors) && $errors !== []) {
            return count($receipts) - count($errors);
        }

        return count($receipts);
    }

    /**
     * Get the queue or return the default.
     */
    public function getQueue(string $queue = null): string
    {
        if (is_string($queue)) {
            return $queue;
        }

        return $this->default;
    }

    /**
     * The underlying MNS queue.
     *
     * @return \Dew\Mns\Versions\V20150606\Queue
     */
    public function getMns()
    {
        return $this->mns;
    }
}
